In [1]:
import pandas as pd

### Attendance Data



Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.



In [2]:
attendance = pd.read_csv('untidy_data/attendance.csv')

In [3]:
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [4]:
attendance.rename(columns={'Unnamed: 0': 'student'}, inplace=True)

In [5]:
attendance

student 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0   Sally          P          T          T          H          P          A   
1    Jane          A          P          T          T          T          T   
2   Billy          A          T          A          A          H          T   
3    John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [6]:
attendance = attendance.melt(id_vars='student', var_name='date', value_name='status')
attendance.head()

student        date status
0   Sally  2018-01-01      P
1    Jane  2018-01-01      A
2   Billy  2018-01-01      A
3    John  2018-01-01      P
4   Sally  2018-01-02      T

In [7]:
attendance.status.unique()

array(['P', 'A', 'T', 'H'], dtype=object)

In [8]:
condition_P = (attendance['status'] == 'P')
attendance.loc[condition_P, 'status']= 1

condition_A = (attendance['status'] == 'A')
attendance.loc[condition_A, 'status']= 0

condition_T = (attendance['status'] == 'T')
attendance.loc[condition_T, 'status']= 0.1

condition_H = (attendance['status'] == 'H')
attendance.loc[condition_H, 'status']= 0.5

In [9]:
attendance['status'] = pd.to_numeric(attendance['status'])

In [10]:
attendance.groupby('student').mean()

status
student        
Billy    0.2250
Jane     0.1875
John     0.7125
Sally    0.3625

### Coffee Levels

- Read the coffee_levels.csv file.
- Transform the data so that each carafe is in it's own column.
- Is this the best shape for the data?


In [11]:
coffee = pd.read_csv('untidy_data/coffee_levels.csv')
coffee.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [12]:
coffee = coffee.pivot(index='hour', columns='coffee_carafe')


In [13]:
coffee.columns = coffee.columns.droplevel()
coffee.columns.name = ''

In [14]:
coffee

x         y         z
hour                              
8     0.816164  0.189297  0.999264
9     0.451018  0.521502  0.915990
10    0.843279  0.023163  0.144928
11    0.335533  0.235529  0.311495
12    0.898291  0.017009  0.771947
13    0.310711  0.997464  0.398520
14    0.507288  0.058361  0.864464
15    0.215043  0.144644  0.436364
16    0.183891  0.544676  0.280621
17    0.391560  0.594126  0.436677

### Cake Recipes



Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.


In [15]:
cake = pd.read_csv('untidy_data/cake_recipes.csv')
cake.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

Tidy the data as necessary.


In [25]:
rnp = cake['recipe:position'].str.split(':', expand=True)
rnp.columns = ['recipe', 'position']
cake = pd.concat([rnp, cake], axis=1).drop(columns=['recipe:position'])

In [26]:
cake

recipe position        225        250        275        300
0      a   bottom  61.738655  53.912627  74.414730  98.786784
1      a      top  51.709751  52.009735  68.576858  50.228470
2      b   bottom  57.095320  61.904369  61.196980  99.248541
3      b      top  82.455004  95.224151  98.594881  58.169349
4      c   bottom  96.470207  52.001358  92.893227  65.473084
5      c      top  71.306308  82.795477  92.098049  53.960273
6      d   bottom  52.799753  58.670419  51.747686  56.183110
7      d      top  96.873178  76.101363  59.571620  50.971626

Which recipe, on average, is the best? recipe b


In [56]:
cake.groupby('recipe').mean().mean(axis=1).idxmax()

'b'

Which oven temperature, on average, produces the best results? 275


In [55]:
cake.mean().idxmax()

'275'

Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees


In [68]:
cake

recipe position        225        250        275        300
0      a   bottom  61.738655  53.912627  74.414730  98.786784
1      a      top  51.709751  52.009735  68.576858  50.228470
2      b   bottom  57.095320  61.904369  61.196980  99.248541
3      b      top  82.455004  95.224151  98.594881  58.169349
4      c   bottom  96.470207  52.001358  92.893227  65.473084
5      c      top  71.306308  82.795477  92.098049  53.960273
6      d   bottom  52.799753  58.670419  51.747686  56.183110
7      d      top  96.873178  76.101363  59.571620  50.971626